### ISSUE TITLE
Cron job executing twice - Leap second issue?

### ISSUE THREAD

**Issue by adgad** [kelektiv/node-cron#180]
We had an issue this morning where our [cron job](https://github.com/Financial-Times/next-daily-email-generator/blob/master/jobs/get-users.js#L65) seemed to execute twice.

This is really just a wild theory, but is it possible that this could have been caused by the leap second? Has anyone else experienced something similar today? 

We're using v1.0.9, running on Heroku.

**Comment by TechnotronicOz** [comment]
@adgad We had the exact same issue where the cron job executed twice this morning.

**Comment by corykrug** [comment]
Happened to us as-well.

**Comment by adgad** [comment]
It doesn't look like it happened again today (after a few deploys/restarts yesteday)

**Comment by moroccan-dude** [comment]
We got the exact same problem. It definitely triggers twice.
We have a cron job that is triggered every night at 4:00 AM. And I have been monitoring the logs  now for 3 nights in a row. The logs are clear and consistent : The job is triggered a first time about 7-8 seconds before 4:00 AM, and then re-triggered at 4:00 AM.
It seems as if this 'double' trigger problem only happens when you schedule the job to run a while after you start the cronjob. In other words, If I set the job to run at 5 PM and start the cronjob at 1 PM, the job only executes once ! But If set the job to run at 04:00 AM and start the cronjob at 1 PM, the job will be triggered twice. That's what I have noticed.  

Here is a snippet of my nodejs file code, nothing really too fancy : 

var initCronJobs = function()
{
	/* init cache cron */
	var cacheJob = new CronJob({
        cronTime: '00 00 04 * * *',
        onTick: function() {
    	       console.log(JSON.stringify( new Date() ) + ' Cache Job Triggered.....');
               // do my stuff
	  },
	  start: false
	});
	cacheJob.start();
        console.log( 'CACHE CRON Job started and set to trigger every day at 4:00 AM');
};

initCronJobs();

Here is a snippet of the logs:
CACHE CRON Job started and set to trigger every day at 4:00 AM
[ 2015-07-01 17:44:18 ] NodeJS web server listening on port 7676
[ 2015-07-02 03:59:53 ] Cache Job Triggered.....
.... job execution
[ 2015-07-02 04:00:00 ] Cache Job Triggered.....
.... job execution

**Comment by Zambonilli** [comment]
Same thing here we had a cron job running every day at 11:15AM US/Central fire twice on 7/1.

**Comment by parkerproject** [comment]
Same here, mine ran twice

**Comment by kgann** [comment]
I've experienced the same issue. `0 6,9 * * 1-5` seems to execute just before the hour and on the hour.

**Comment by DmitryEfimenko** [comment]
I was able to reproduce this issue consistently and I think I figured out exactly the condition for this issue:
Say the job is to be run every minute (* * * * *).
If you start this job in the second half of the interval (so say at 50th second of the current minute) - the job will execute multiple times. It seems that the closer you get to the end of the interval - the more times (exponentially) the job will execute.

**Comment by DmitryEfimenko** [comment]
any luck with this?

**Comment by dp-ck** [comment]
I'm having the same issue: cron job is always executed twice.

**Comment by mattklein999** [comment]
+1

**Comment by DmitryEfimenko** [comment]
is this project dead?

**Comment by ncb000gt** [comment]
Sorry. The project isn't dead but I'm extremely busy so don't have much time to put towards issues. I'm more than willing to add other contributors or to merge pull requests. But, in the past nobody has seemed very interested in being a committer.

**Comment by avingochea** [comment]
+1 Same issue here. Fires at 11:59am and 12:00pm. Wish I could help, but I wouldn't know where to start!

**Comment by aliuz1** [comment]
Same issue here, it's set to run at the 59th minute of every hour, but sometimes it runs twice, once at the 58th minute and once at the 59th. 
It would be amazing if someone could take a look at this.

**Comment by Grayda** [comment]
I just experienced the same issue. I'm having a glance through the source to see if I can find out what's going on, but one solution I saw involved doing your own checks to see if the job has run.

One workaround that came to mind, was to use something like [cron-parser][1] to get the human-readable date, then do a compare (for example, in pseudo-code: `if(timeJobShouldStart == 1:14pm && currentTime == 1:14pm) { doYourStuffHere() }`

Not a pretty workaround, but at least it'll double-check to ensure important jobs don't run twice. 

I'll do a PR if I can find any bugs, as I'm using this for an open source school bell system, and firing a bell only once is super important :smile:

[1]: https://www.npmjs.com/package/cron-parser

**Comment by ncb000gt** [comment]
@adgad This is entirely possible. I know there are some outstanding DST issues right now. So, this wouldn't surprise me. =\ Unfortunately it's a complicated issue. I'll try to figure something out for these cases.

@DmitryEfimenko What version of node are you on? I just put a test in (taken from your example) and I cannot reproduce the issue.

**Comment by DmitryEfimenko** [comment]
my package.json says it's 1.0.9 (which I assume is latest?). However, I noticed that it looks different that what is on the Github. I comparent the contents of local cron.js to the Github cron.js. They are a bit different. The one on Github has function CronJob accepting a new parameter `runOnInit` and couple other changes.
I'm not sure if that makes a difference, but I could try updating to the real latest and testing my code again.

**Comment by ncb000gt** [comment]
Well, I did a bunch of stuff today on the train. I haven't released a new
version yet. Just trying to get caught up on stuff I've been meaning to. ;)

**Comment by DmitryEfimenko** [comment]
@ncb000gt hey, I created a failing runnable test on Cloud9 that you can [see for yourself](https://ide.c9.io/cactusss/test).
I made it public, but I think you'll need to create an account there if you don't have one.
When you open the project, just execute `gulp` in the command line

**Comment by AngelMarquez** [comment]
In case it helps anyone: I was able to work around this bug this bug from happening by importing lodash and using its 'throttle' function. https://lodash.com/docs#throttle

I just wrap every function to be called in `_.throttle()

Like so:
```
  new cronJob('*/15  *  * * *   *', _.throttle(myDoubleFiringFunc, 2000), null, true, tz)
```
Just make sure that your throttle value is sufficiently large for your use. The 2 second throttle above works perfectly for me.

I know it's clunky. But it works.

**Comment by avingochea** [comment]
Great find. Thanks a lot Angel!

**Comment by arsinawaz** [comment]
@ncb000gt any update on the issue?

**Comment by arsinawaz** [comment]
@ncb000gt  any update on the issue? i am facing the same problem

**Comment by pgetta** [comment]
yes, the same problem here.

**Comment by khejing** [comment]
@AngelMarquez I even use "new cronJob('*/15  *  * * *   *', _.throttle(myDoubleFiringFunc, 120000), null, true, tz)", but still have this problem. It just execute the second time after 2 minutes later!

**Comment by samzilverberg** [comment]
@khejing i've also encountered this problem and have finally decided to just use another cron lib

**Comment by arsinawaz** [comment]
I faced the same problem and found that it was due to node running on two clusters in my case. you can use one cluster to execute the job.
`var cluster = require('cluster');
if(cluster.worker.id == 1){
 // run job
}`

**Comment by shreyas-sja** [comment]
+1, Found that cron triggered twice after 10 minutes of difference

**Comment by AngelMarquez** [comment]
Sorry guys, the lodash fix seemed to work, but it eventually started doing as you all reported and just firing after the window expired. I'm at a loss as to why that is.

**Comment by Zambonilli** [comment]
This happened to us again in production.  However, I think we have a beat on why it's happening.  We updated our NTP servers last week in production.  The server running cron was ~1.5 seconds slow from the NTP server.  However, the process that had the cron job running did not have the process restarted.  So, there was an outstanding frame on the stack for the cron job with a setTimeout that was computed with the slow clock.  The clock sped up about 1.5 seconds underneath the setTimeout calculated value.  Therefore, the calling of self.start() in the callbackWrapper function computed the next time to fire to be in 1.5 seconds even though the cron mask was for once a week.

I'm thinking the fix here should be to either hook into the system's clock methods and recompute outstanding timeouts or store a last callback timestamp and recompute the current time/time until and make sure it makes sense.

Also, I'm trying to figure out a way to test this.  Unfortunately, from a couple hours of poking at it there might not be a way to simulate clock changes w/o setTimeout callbacks being fired in sinon.

**Comment by brunocasado** [comment]
apparently especify a second like 05 "solve the problem"

**Comment by ChrisCinelli** [comment]
This should fix this problem: https://github.com/kelektiv/node-cron/pull/259

**Comment by flaw600** [comment]
This is still going on

**Comment by vipulpandeykv** [comment]
if you all are using cluster then make another node app.js or server.js

**Comment by srikantk** [comment]
It is still happening to me. I modified the cron posted by @AngelMarquez
from:
 ` new cronJob('*/15  *  * * *   *', _.throttle(myDoubleFiringFunc, 2000), null, true, tz)`
to:
  `  new cronJob('*/15  *  * * *   *', _.throttle(myDoubleFiringFunc, 2000, {leading:true, trailing:false}), null, true, tz)`

Seems to work for now.

**Comment by Sundin** [comment]
I have the same problem.

Trying @srikantks workaround, seems to work so far.

```
const job = new CronJob({
    cronTime: '*/15 * * * * *',
    onTick: _.throttle(functionToRun, 2000, { leading: true, trailing: false }),
    start: true,
    timeZone: 'Europe/Stockholm'
});
```

**Comment by shishir99111** [comment]
Finally, got the solution to resolve it in my case. 
As in my case, I have specifically make sure that CronJob should get started after the mongo connect event on Nodejs Process startup. Just like this..

``
require('./config/mongoose').initialize(() => {
   const cronJob = require('./cronJob');
   cronJob.initialize();
});
``
Also, my mongo has multiple instances with Replication Election Protocol. So, whenever my primary mongo gets disconnected it gets connection back to the secondary server(electing it as primary) so, it again listens to the mongo connect event (without server getting restart), that leads to the multiple CRON Jobs running getting initiated. So, my solution was:

``
require('./config/mongoose').init();
require('./worker').init();
``

**Comment by ncb000gt** [comment]
So, I know this ticket is crazy old and yall seem to still be seeing the behavior. I've merged in a few prs and released a new version. I've also added a handful more testing for different cases and scenarios. Is anyone still seeing this in isolation? I haven't seen this happening recently, but it could be isolated to my machines/use cases.

Let me know. Thanks. Also, sorry for the age of this. Apparently all my GH email was going to a label that I wasn't being notified of. =\ That's been fixed at least...

**Comment by ncb000gt** [comment]
Closing due to age. If this is still an issue with the latest (1.4.1 as of this comment) then we can open a new issue.

**Comment by galoberlyn** [comment]
I just did a 24 hour testing regarding this issue and I thought this might help with some of you who's experiencing the 2 time execution

It seems to me that schedule every 59 minutes triggers twice in both 59 and 00 minute mark.

`const syncProspectToCoreUser = cron.schedule("*/59 * * * *", () => {
    var d = new Date();
    var s = d.getSeconds();
    var m = d.getMinutes();
    var mm = d.getMilliseconds();

    
    myConsole.log('Trigged @ minute', m);
    myConsole.log('Trigged @ seconds', s);
    myConsole.log('Trigged @ milliseconds', mm);
    myConsole.log('==================================================================');
    console.log('triggered');
});`

Output.txt
![image](https://user-images.githubusercontent.com/25279611/115664471-733f6380-a374-11eb-8757-4c701a166941.png)


Somehow, when changing the schedule to exactly 1 hour, the issue has been fixed. Not sure if this is a wrong implementation on my end but it's now working on my end. 😊 

`const syncProspectToCoreUser = cron.schedule("0 * * * *", () => {
    var d = new Date();
    var s = d.getSeconds();
    var m = d.getMinutes();
    var mm = d.getMilliseconds();

    
    myConsole.log('Trigged @ minute', m);
    myConsole.log('Trigged @ seconds', s);
    myConsole.log('Trigged @ milliseconds', mm);
    myConsole.log('==================================================================');
    console.log('triggered');
});`
![image](https://user-images.githubusercontent.com/25279611/115664608-a3870200-a374-11eb-867a-caf3b262a6ac.png)

**Comment by thekennysong** [comment]
I just ran into this issue. Will look into the lodash solution, not sure why it triggered twice as it’s the first time I’ve ran into this issue and I started using this a couple months ago. Love the project tho!

**Comment by soorur** [comment]
@thekennysong did the lodash workaround fix the situation for you?
Happened to us in production as well - running on a single instance and does not happen all the time.

Cron job is initialized once when service is registered

Timestamps:
1: 2021-08-05T15:00:07.752193632Z
2: 2021-08-05T15:00:07.951931042Z

**Comment by thekennysong** [comment]
@soorur it turned out it wasn't the library but that I had 2 dynos running for Express app. Didn't cross my mind that it would trigger 2 cron jobs haha. Regardless I did wrap my cron job in a lodash throttle just to be extra safe.

**Comment by noukaza** [comment]
This is still going on

**Comment by sinmichael** [comment]
> apparently especify a second like 05 "solve the problem"

This worked for me.

**Comment by lukaszgebel** [comment]
I am getting the issue it happens only in production. I thiks it is related to restartig the server after changes.

**Comment by intcreator** [comment]
seeing as this is a 10 year old issue I doubt it's the same problem. if you would like to provide more details and a [minimum complete reproducible example](https://stackoverflow.com/help/minimal-reproducible-example), please open a new issue and follow the issue template to do so
